### PR TITLE
Don't count the toolbars and main viewport in the window limit.

### PR DESCRIPTION
**Pull request by Wirlie**
This is a clean contribution of #4115 (Started by @mrtnptrs). Also I've added the "remaining" solution for this problem.

**Note:** I think that this is not the real solution, this is a fast solution and it works. i'll see if I can improve this later.

If @wolfreak99 @wolfreak100 have a better solution then, this can be marked as closed.

### CODE REVIEW

**Comment by janisozaur** [comment]
@Wirlie see https://github.com/OpenRCT2/OpenRCT2/wiki/Commit-Messages

**Comment by mrtnptrs** [comment]
I can confirm that it works correctly now! :) Although, instead of -3 and +3 workarounds, it just shouldn't count the toolbars and main viewport as a window, in my opinion. But the solution for that could be pretty complicated, so, maybe this workaround is a good solution. But, as I'm not a developer, you can descide it best :)

File: src/interface/window.c [hunk 1]
```diff
@@ -402,7 +402,7 @@ void window_set_window_limit(int value)
 rct_window *window_create(int x, int y, int width, int height, rct_window_event_list *event_handlers, rct_windowclass cls, uint16 flags)
 {
 	// Check if there are any window slots left
-	if (RCT2_NEW_WINDOW >= &(g_window_list[gConfigGeneral.window_limit])) {
+	if ((RCT2_NEW_WINDOW - 3) > &(g_window_list[gConfigGeneral.window_limit])) {
```
Comment (mrtnptrs) [inline src/interface/window.c]: Just curious: why are you using ">" now instead of ">="?

**Comment by Wirlie** [comment]
Thanks @janisozaur. It is better?

File: src/interface/window.c (hunk 1 above)
Comment (Wirlie) [inline src/interface/window.c]: For example, if the window limit is 8, with `>=` only you can have 7 windows at the same time, but with `>`, you can have 8 windows.

Is curious xP, probably a mistake.

**Comment by mrtnptrs** [comment]
@Wirlie As I'm a giant loser when it is about coding (at the moment, want to improve it in the comming months), could you take a look at the above comment from @wolfreak99 ?

**Comment by mrtnptrs** [comment]
@wolfreak99 But isn't it a bit strange that OpenRCT2 counts tooltips, toolbars and even the main viewport as a window? (sorry, I know, same question I asked previously, but I think I never got an answer on that one) Not that I'm a code expert (far from it), but can't you add a flag to the actual windows to recognize them or add a flag to the toolbars, main viewport and the tooltips to NOT recognize them as a window? Or would introducing a new flag break compatibility with RCT2 vanilla?

**Comment by Wirlie** [comment]
@mrtnptrs I've already done it :) thanks. Can you test my new changes? 
@wolfreak99 Can you take a look to my new changes?

**Comment by mrtnptrs** [comment]
@Wirlie looking at it now!

**Comment by Wirlie** [comment]
@mrtnptrs done, thanks for your testing :)

**Comment by mrtnptrs** [comment]
@Wirlie Testing it now again.

**Comment by mrtnptrs** [comment]
Ok, I see a few issues here (WIP!):
-It can take a long time (up to 15 seconds or so) till a window closes when you reached the window limit; So, when the limit is set to 9, and you open windows number 10, it can take a long time before an other window closes to match the window limit. Seems like the check happens too late, but it actually should get checked every time a new window gets opened.
-Sometimes a viewport window of a guest (or another window?) just closes without reaching the window limit. I have no reliable way to reproduce this and will try to find specific steps for you to reproduce it, so you can fix it.

**Comment by Wirlie** [comment]
@mrtnptrs I think that these problem has to do with the code that @wolfreak99 has implemented. In that case, @wolfreak99 should take a look at these problems.

Additionally, I know that this is a bad solution, because this solves the problem using "magic values", instead of solving the problem that has to do with the 3 persistent windows. That you've said before is the real solution:
> it just shouldn't count the toolbars and main viewport as a window

**Comment by mrtnptrs** [comment]
@wolfreak99 Then we will wait a few days for him to return :)

**Comment by mrtnptrs** [comment]
@Wirlie @wolfreak99 And we have to take into account that tooltips also count as windows; when you set the limit to 8, open 8 windows and then you hover you cursor over the toolbar, one window will close. Maybe @wolfreak99 could make a new PR with the nice and clean solution: to not let the game count toolbars, the main viewport en tooltips as windows? (trough a flag or something?)

**Comment by wolfreak99** [comment]
@mrtnptrs 
Yes the original "counted the tooltips as windows" because the tooltips, etc, ARE windows, they just don't have that titlebar at the top that makes it appear as if it's a window.
https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/windows/tooltip.c#L124
This is the file for the tooltip, look at the line i linked to. once the game is fully reverse engineered, then the tooltips, toolbars, etc, can be seperated into it's own unique lists, but as for now, there are still calls to the main game.
The game stores a list that can hold up to a maximum of values (this is called an array, look it up if you're not famliar with them). In the games update loop, it runs through each window slot and sees if it has something in it, and will then call that items update functions, which is essential to updating all the windows at once.

Think of it this way. you have a shelf for nintendo cartridges. all nintendo cartridges are generally a specific width, so this shelf can hold a maximum of 10 cartridges per shelf. You are theoretically wanting that shelf of 6 cartridges to not count the 7th cartridge (because, say, it's an application more than it is a game), so you take that cartridges label off and put it on the shelf. to your surprise, there is still only room for 3 more slots after ripping the label off.

Does this now make sense to you?

**Comment by wolfreak99** [comment]
https://github.com/OpenRCT2/OpenRCT2/pull/4144 i have just uploaded a fix that kinda works in the way that this was attempted, but it addresses it more properly.

**Comment by Wirlie** [comment]
Thanks  @wolfreak99 , thats is more clean, i'll test your PR tomorrow.

**Pull request closed without merging**